### DISCUSSION TITLE
Increase latitude and longitude decimal precision of MPR output

### DISCUSSION THREAD

**Discussion by wangrenz** [dtcenter/METplus#1548] [Configuration] [answered]
Point_stat OBS_LAT OBS_LON have two decimals of MPR output now.

like this:
```
OBS_LAT OBS_LON
42.77   16.9
47.6    12.65
59.25 -105.84
33.5   114.56
```

How to increase decimal to tree or more.

Thanks.

**Comment by JohnHalleyGotway** [comment, marked as answer]
@wangrenz that's a great question. There is one configurable option in MET related to precision. It's named "output_precision", shows up in the [ConfigConstants](https://github.com/dtcenter/MET/blob/71f3d427506e9bea98a813a0ffd30039071e9aac/met/data/config/ConfigConstants#L16) configuration file, and is described in [this section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=output_precision#output-precision) of the MET User's Guide.

Whenever a MET tool is run, it first reads the ConfigConstants config file, then the default configuration file for that tool, and lastly, the config file specified by the user on the command line. So adding "output_precision = 7;" (or whatever) to the config file you're using will override the default value of 5 read from ConfigConstants.

It looks like this is NOT currently configurable directly via the METplus wrappers. But you could always specify it using [POINT_STAT_MET_CONFIG_OVERRIDES](https://metplus.readthedocs.io/en/latest/Users_Guide/glossary.html?highlight=output_precision#term-POINT_STAT_MET_CONFIG_OVERRIDES):
```
POINT_STAT_MET_CONFIG_OVERRIDES = output_precision = 7;
```

However, this may not actually change any of the output lat/lon values you're seeing. MET can only output the lat/lon precision that it finds in the input. If the input only has 2 decimal places, then that's all the output can have.

This [line of code](https://github.com/dtcenter/MET/blob/71f3d427506e9bea98a813a0ffd30039071e9aac/met/src/tools/core/point_stat/point_stat.cc#L506) in Point-Stat uses output_precision to setup the output AsciiTable object, which is used to buffer the data that Point-Stat writes. The AsciiTable class writes any numbers less than -1 or greater than 1 using the %g formatting option for [printf](https://www.cplusplus.com/reference/cstdio/printf/). Put simply, you'll only see more than 2 decimal points of the precision if they actually exist in the input.

I ran a quick test to make sure this all works the way I expect.
1. Modify an ascii obs file by adding lots more (fake) precision to the input lat/lon values:
```
cat data/sample_obs/ascii/sample_ascii_obs.txt | awk '{print $1,$2,$3,$4"0123456789",$5"0123456789",$6,$7,$8,$9,$10,$11}' > tmp.txt
mv tmp.txt data/sample_obs/ascii/sample_ascii_obs.txt
```

2. Make sure it worked:
```
head -2 data/sample_obs/ascii/sample_ascii_obs.txt | awk '{print $4, $5}'
35.030123456789 -106.620123456789
35.030123456789 -106.620123456789
```

3. Re-run "make test" and check the MPR output from Point-Stat:
```
40.15012  -89.33012
```
I definitely see the extra precision specified there. I did also try increasing output_precision, and did see more decimal places in the output as a result.

Note though that the if you're running ascii2nc, the output_precision setting (from ConfigConstants) is also used there. If you'd like to override that default, you'd need to do so by specifying a config file for it on the command line.

I can see how adding support for a MET_OUTPUT_PRECISION environment variable would be useful. If set, its value would override anything read from a config file. But that option does not currently exist. Do you think you'd find that option useful?

Hope this helps clarify.

**Reply by JohnHalleyGotway** [reply]
@wangrenz I wanted to followup on this discussion. Does that answer your question? If so, feel free to click the "Mark as answer" button.
